### ISSUE TITLE
Hunyuan3D-2mv

### ISSUE THREAD

**Issue by TristanMorelle** [kijai/ComfyUI-Hunyuan3DWrapper#95]
@kijai 

Please add support for Hunyuan3D-2mv
https://huggingface.co/tencent/Hunyuan3D-2mv

or if it would work please provide a workflow.

Ty

**Comment by kijai** [comment]
It needs some work to add support for the model, looks like it can be added to this repo though yes.

**Comment by AdamGman** [comment]
The new MV one is high res, but also has MV fast. and also is new Mini and mini-fast! MV loads in this plugin now, but it is probably just brute force it without optimization

**Comment by TristanMorelle** [comment]
I was hoping a unify as batch approach would cover it , but i don't think it's working 😆 
ps Noob here


```
class Hy3DGenerateMeshMultiview:
    @classmethod
    def INPUT_TYPES(s):
        return {
            "required": {
                "pipeline": ("HY3DMODEL",),
                "image": ("IMAGE", ),
                "guidance_scale": ("FLOAT", {"default": 5.5, "min": 0.0, "max": 100.0, "step": 0.01}),
                "steps": ("INT", {"default": 30, "min": 1}),
                "seed": ("INT", {"default": 0, "min": 0, "max": 0xffffffffffffffff}),
            },
            "optional": {
                "mask": ("MASK", ),
            }
        }

    RETURN_TYPES = ("HY3DLATENT",)
    RETURN_NAMES = ("latents",)
    FUNCTION = "process"
    CATEGORY = "Hunyuan3DWrapper"
    INPUT_IS_LIST = (False,True,False,False,False,True)

    def process(self, pipeline, image, steps, guidance_scale, seed, mask=None):

        def unlist(var):
            if hasattr(var, '__len__') and len(var) == 1 and not isinstance(var[0], torch.Tensor):
                var = var[0]
            return var

        def wrap_tensor(tensor):
            if isinstance(tensor, torch.Tensor):
                tensor = [tensor]
            return tensor
        
        # unwrap lists from INPUT_IS_LIST
        pipeline = unlist(pipeline)
        image = unlist(image)
        steps = unlist(steps)
        guidance_scale = unlist(guidance_scale)
        seed = unlist(seed)
        mask = unlist(mask)
        
        # comfirm images are in list
        mask = wrap_tensor(mask)
        image = wrap_tensor(image)

        mm.unload_all_models()
        mm.soft_empty_cache()

        device = mm.get_torch_device()
        offload_device = mm.unet_offload_device()

        images = []
        masks = []

        if mask is None:
            mask = [None for i in image]

        if type(image) is not type(mask):
            raise RuntimeError('mixed types not supported:', type(image),len(image),type(mask),len(mask))

        if not isinstance(image, torch.Tensor) and not (isinstance(mask, torch.Tensor) or mask is None):
            if len(image) != len(mask):
                raise RuntimeError('image and mask does not have the same length:', image,mask)

        for ima_i, mask_i in zip(image, mask):
            ima_i = ima_i.permute(0, 3, 1, 2).to(device)
            ima_i = ima_i * 2 - 1
            images.append(ima_i)

            if mask_i is not None:
                mask_dims = len(mask_i.shape)
                if mask_dims == 2: # (H, W) -> (1, 1, H, W)
                    mask_i = mask_i.unsqueeze(0).unsqueeze(0).to(device)
                elif mask_dims == 3: # (B, H, W) -> (B, 1, H, W)
                    mask_i = mask_i.unsqueeze(1).to(device)
                elif mask_dims == 4: # (B, C, H, W) -> (B, 1, H, W) - Assuming we need 1 channel
                    if mask_i.shape[1] > 1:
                        mask_i = mask_i[:, 0:1, :, :].to(device)
                    else:
                        mask_i = mask_i.to(device)
                else:
                    raise ValueError(f"Unexpected mask shape: {mask_i.shape}")

                if mask_i.shape[2] != ima_i.shape[2] or mask_i.shape[3] != ima_i.shape[3]:
                    mask_i = F.interpolate(mask_i, size=(ima_i.shape[2], ima_i.shape[3]), mode='nearest')
                masks.append(mask_i)
            else:
                masks.append(None)

        images_tensor = torch.cat(images, dim=0)
        masks_tensor = torch.cat([m for m in masks if m is not None], dim=0) if any(m is not None for m in masks) else None

        print (f"images => {images_tensor.shape}")
        print (f"masks => {masks_tensor.shape if masks_tensor is not None else None}")

        pipeline.to(device)

        try:
            torch.cuda.reset_peak_memory_stats(device)
        except:
            pass

        latents = pipeline(
            image=images_tensor,
            mask=masks_tensor,
            num_inference_steps=steps,
            guidance_scale=guidance_scale,
            generator=torch.manual_seed(seed))

        print_memory(device)
        try:
            torch.cuda.reset_peak_memory_stats(device)
        except:
            pass

        pipeline.to(offload_device)

        return (latents, )
```

**Comment by kijai** [comment]
Nodes have been updated to support the MV models, and example workflow too. I'm not entirely sure if we should be sending multiple images to the multiview texturing node, it accepts them but even the original code doesn't use anything to assign the references to specific camera views for texturing.

**Comment by AdamGman** [comment]
You are amazing for doing this so fast. I have a feeling it does want to accept multiple images and this custom nodes would workflow that.  https://github.com/huanngzh/MV-Adapter?tab=readme-ov-file#partial-image--geometry-to-multiview

**Comment by kijai** [comment]
> You are amazing for doing this so fast. I have a feeling it does want to accept multiple images and this custom nodes would workflow that. But maybe it doesn't actually need them? https://github.com/huanngzh/MV-Adapter?tab=readme-ov-file#partial-image--geometry-to-multiview

The multiple images to the mesh generation absolutely works and pretty well too. For the texture generation part I'm still unsure.

**Comment by AdamGman** [comment]
Dumb question but do you know if its possible to get this MV adapter into your workflow? the issue is it creates all images into 1 image node and need to separate out each angle to its own image. Or do you have another way of doing this with t2i 


![Image](https://github.com/user-attachments/assets/901c4819-a8f5-48b5-a25c-f6e49392febe)

**Comment by jakechai** [comment]
@kijai Thanks for the great work. Are there any tricks when setting the MVs? I checked the MV generation mesh workflow but got these, switching the order does not help. The result is not like in the official demo. (I also found that the textures are not using the input MVs.)
![Image](https://github.com/user-attachments/assets/e7aae619-d42a-4122-9422-428b59ee3b63)

![Image](https://github.com/user-attachments/assets/5eccb163-a204-463e-9131-f08d95cdecd1)

![Image](https://github.com/user-attachments/assets/23d76d4e-a361-467f-95f2-5e5bb2f162e5)

![Image](https://github.com/user-attachments/assets/2378fc24-0fbb-4e40-a8f6-e411be98baf6)

![Image](https://github.com/user-attachments/assets/13968a20-24d4-454f-93fc-9f5088b08dba)

![Image](https://github.com/user-attachments/assets/85480664-d43c-4e75-aa8b-fb058efb9c64)

Scaling down the subject to a 0.85 ratio makes it better but not ideal.

![Image](https://github.com/user-attachments/assets/d85b4fc6-0f93-4f3a-8444-0bf96dd12df6)

**Comment by AdamGman** [comment]
> [@kijai](https://github.com/kijai) Thanks for the great work. Are there any tricks when setting the MVs? I checked the MV generation mesh workflow but got these, switching the order does not help. The result is not like in the official demo. (I also found that the textures are not using the input MVs in the official demo.) ![Image](https://github.com/user-attachments/assets/e7aae619-d42a-4122-9422-428b59ee3b63)
> 
> ![Image](https://github.com/user-attachments/assets/5eccb163-a204-463e-9131-f08d95cdecd1)
> 
> ![Image](https://github.com/user-attachments/assets/23d76d4e-a361-467f-95f2-5e5bb2f162e5)
> 
> ![Image](https://github.com/user-attachments/assets/2378fc24-0fbb-4e40-a8f6-e411be98baf6)
> 
> ![Image](https://github.com/user-attachments/assets/13968a20-24d4-454f-93fc-9f5088b08dba)
> 
> ![Image](https://github.com/user-attachments/assets/85480664-d43c-4e75-aa8b-fb058efb9c64)
> 
> Scaling down the subject to a 0.85 ratio makes it better but not ideal.
> 
> ![Image](https://github.com/user-attachments/assets/d85b4fc6-0f93-4f3a-8444-0bf96dd12df6)

I got this working with this MV adapter plugin.

https://github.com/huanngzh/ComfyUI-MVAdapter

you make it do 6 sides, then use "Image Batch Splitter" from Inspire pack custom node or (other image splitter), then you can put each image exactly on respective sides from this workflow easier. I also recommend to do it 1024 and switch the resizing to after background remove to save on mask quality, and the Delight group can be totally bypassed as MV adapter does this already.

**Comment by jakechai** [comment]
Great, thanks a lot.

**Comment by kijai** [comment]
> @kijai Thanks for the great work. Are there any tricks when setting the MVs? I checked the MV generation mesh workflow but got these, switching the order does not help. The result is not like in the official demo. (I also found that the textures are not using the input MVs in the official demo.)
> ![Image](https://github.com/user-attachments/assets/e7aae619-d42a-4122-9422-428b59ee3b63)
> 
> ![Image](https://github.com/user-attachments/assets/5eccb163-a204-463e-9131-f08d95cdecd1)
> 
> ![Image](https://github.com/user-attachments/assets/23d76d4e-a361-467f-95f2-5e5bb2f162e5)
> 
> ![Image](https://github.com/user-attachments/assets/2378fc24-0fbb-4e40-a8f6-e411be98baf6)
> 
> ![Image](https://github.com/user-attachments/assets/13968a20-24d4-454f-93fc-9f5088b08dba)
> 
> ![Image](https://github.com/user-attachments/assets/85480664-d43c-4e75-aa8b-fb058efb9c64)
> 
> Scaling down the subject to a 0.85 ratio makes it better but not ideal.
> 
> ![Image](https://github.com/user-attachments/assets/d85b4fc6-0f93-4f3a-8444-0bf96dd12df6)

That kinda looks like what you'd get when. It using the MV model? No tricks besides setting the right image to right view slot. 

About the texture gen part I don't know, it can accept multiple images or just one, I don't know which is better.

**Comment by jakechai** [comment]
The MVs I got were generated using Hunyuan3D. What's weird is that the generation looks fine on the official demo using the same MVs.

**Comment by AdamGman** [comment]
Here is my beginning to your workflow for T23D if anyone needs help. Maybe I did it too complex but works perfect.

![Image](https://github.com/user-attachments/assets/416cadff-64a5-474d-b8fe-50322a402711)

**Comment by kijai** [comment]
> The MVs I got were generated using Hunyuan3D. What's weird is that the generation looks fine on the official demo using the same MVs.

I mean did you use the multiview model? Because those generations look exactly what you get when not using multiview model with multiple inputs.

**Comment by kijai** [comment]
Here's the result with the example workflow with the multiview model, just from screenshots of your images:

https://github.com/user-attachments/assets/cb429740-6b91-4b86-998a-ce9e85266d55

**Comment by kijai** [comment]
> > I mean did you use the multiview model? Because those generations look exactly what you get when not using multiview model with multiple inputs.
> 
> Oh yes, my bad, I forgot to select the mv model, sorry for wasting your time.

No worries! It's interesting the old models even work at all with multiple inputs like that.
